### PR TITLE
Fix duplicated FldSeq in block morphing

### PR DESCRIPTION
**Pull request by SingleAccretion**
Reusing the address for the first field is problematic as the first field is likely to be at a zero offset, thus a zero-offset field sequence will be attached to it, and subsequent copies of the same tree will pick it up, which is incorrect and can result in silent bad codegen.

Fix by reusing the address for the last field instead.

```diff
 [000015] -A-X-+------              *  COMMA     void
-[000008] -A-X--------              +--*  ASG       long
-[000006] *--X---N----              |  +--*  IND       long
-[000000] -----+------              |  |  \--*  LCL_VAR   byref  V00 arg0          Zero Fseq[FirstLngValue]
-[000007] ------------              |  \--*  LCL_VAR   long   V03 tmp1
+[000009] -A-X--------              +--*  ASG       long
+[000007] *--X---N----              |  +--*  IND       long
+[000006] -----+------              |  |  \--*  LCL_VAR   byref  V00 arg0          Zero Fseq[FirstLngValue]
+[000008] ------------              |  \--*  LCL_VAR   long   V03 tmp1
 [000014] -A-X--------              \--*  ASG       long
 [000012] *--X---N----                 +--*  IND       long
 [000011] ------------                 |  \--*  ADD       byref
-[000009] -----+------                 |     +--*  LCL_VAR   byref  V00 arg0          Zero Fseq[FirstLngValue]
+[000000] -----+------                 |     +--*  LCL_VAR   byref  V00 arg0
 [000010] ------------                 |     \--*  CNS_INT   long   8 Fseq[SecondLngValue]
 [000013] ------------                 \--*  LCL_VAR   long   V04 tmp2
```

Part of #58312.

Just [one](https://dev.azure.com/dnceng/public/_build/results?buildId=1509076&view=ms.vss-build-web.run-extensions-tab) method with diffs: "better" sequences unblocked quite a few CSEs.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
Will request the `Linux_musl arm` be re-run: it is a Regex test timeout on an ARM32 in Debug.

@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
> Will request the Linux_musl arm job be re-run

Rerunning.

File: src/coreclr/jit/morphblock.cpp [hunk 1]
```diff
@@ -1259,9 +1259,9 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
                     }
                     else
                     {
-                        if (i == 0)
+                        if (i == (fieldCnt - 1))
                         {
-                            // Use the orginal m_dstAddr tree when i == 0
+                            // Reuse the orginal m_dstAddr tree for the last field.
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morphblock.cpp:1264]: Should we assert that we're not reusing an address with a zero offset field seq?

File: src/tests/JIT/Directed/StructPromote/FldSeqsInPromotedCpBlk.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+class FldSeqsInPromotedCpBlk
```
Comment (AndyAyersMS) [inline src/tests/JIT/Directed/StructPromote/FldSeqsInPromotedCpBlk.cs:6]: Can you add a comment linking this test to the PR and/or describing the problem this test is supposed to cover?

File: src/coreclr/jit/morphblock.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/morphblock.cpp:1264]: Hmm, it is legal for the address to contain a zero-offset sequence before the morphing, so I suppose asserting here would require:

1) Getting the "base" sequence (essentially perform the reverse of what `fgAddFieldSeqForZeroOffset` does).
2) After morphing, loop over all the created address trees and assert that we only added one field to the "base", and all of them are distinct.

I am not sure it is worth the amount of code required. Once #58312 is finished, we will have asserts in VN that would have caught this case.

**Comment by AndyAyersMS** [comment]
Native AOT build (a newly added thing, see #62833) is failing. I'm not familiar with this yet, so can't say what's up.

It is running tests during the build step which means test failures don't get summarized the same way as for the other CI legs, so I have no idea if this failure is happening elsewhere too.

cc @MichalStrehovsky @dotnet/jit-contrib 

```
      nativeaot/SmokeTests/BasicThreading/BasicThreading/BasicThreading.sh [FAIL]
```

**Comment by MichalStrehovsky** [comment]
Looks like the FinalizeTest also recently failed in runtimelab: https://github.com/dotnet/runtimelab/pull/1702#issuecomment-962491698. The failure is probably unrelated.

I've submitted https://github.com/dotnet/runtime/pull/62924 to increase the reliability of the test.

File: src/coreclr/jit/morphblock.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/morphblock.cpp:1264]: I'm not a fan of how we handle the zero offset case today. We've had a number of bugs in this area.

Ok by me if it is too hard to check -- just wanted to see if you had any ideas.

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**